src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java

```diff
@@ -1948,7 +1950,7 @@ private OutputStream getOutputStream(Entry e) throws IOException {
             e.file = getTempPathForEntry(null);
             os = Files.newOutputStream(e.file, WRITE);
         } else {
-            os = new ByteArrayOutputStream((e.size > 0)? (int)e.size : 8192);
+            os = new ByteArrayOutputStream((e.size > 0 && e.size <= MAX_ARRAY_SIZE)? (int)e.size : 8192);
         }
```

Comment:
The proposed change will address the specific issue shown in the bug.   As Alan points out, there could be an issue if the deflated size is > 2gb.  It would be good to look into that as part of your proposed fix.